### ISSUE TITLE
Client / broker connection performance

### ISSUE THREAD

**Issue by aronweiler** [chkr1011/MQTTnet#155]
Hello!  I just started using this library recently, and I have some questions... 

**1. Connection performance**
I created a simple client/server test in order to look at the performance of MQTTnet, and I noticed that the connection acceptance rate is very slow (using TCP).  I was wondering why this is, and is there something that I am missing?  When I connect 10 clients, and it takes roughly 10 seconds for all of the connections to be fully realized.  This seems very strange (and slow) to me.  I would imagine that the connections should be established almost instantaneously for that small a number of clients.

**2. Number of possible connections**
When I create ~50 clients in my loop (see code), the connection process takes even longer than described above, and very often I can't get more than about 40 clients connected at all.  Is this because I am using the client in a way that it's not intended?  Such as running so many clients from the same machine?

As to why I am looking at this library... 
I have a project that requires me to simulate thousands of MQTT clients (running on the same machine) connecting to a server, and the performance requirements are rather stringent- the clients (10,000) should all be able to connect to the server within seconds, and be ready to send/receive.

I've attached the code for my simple server and client test (pretty much a copy/paste from the wiki), and I'd appreciate any help in understanding why I am seeing these performance issues.

Thanks so much!

Code: [MQTTSpike.zip](https://github.com/chkr1011/MQTTnet/files/1613435/MQTTSpike.zip)

**Comment by slacerte** [comment]
In the same context as above (only 10 clients) I see a significant time gap between the client sending (TX) the MqttConnectPacket and the client receiving (RX) the MqttConnAckPacket.  Note that the log output containing [MQTTnet] namespaces is produced by wiring a handler to `MqttNetTrace.TraceMessagePublished`

[03:03:12.657 - Waiting to connect client: 001ea3ff-2ec4-40ec-8457-20dae3018339...
[03:03:12.710 - Connection of all 1 clients has been initiated...
[03:03:13.722] [4] [MQTTnet.Core.Client.MqttClient] [Information]: Start receiving packets.
**[03:03:13.841]** [3] [MQTTnet.Core.Adapter.MqttChannelCommunicationAdapter] [Information]: **TX >>> MqttConnectPacket**: [ClientId=001ea3ff-2ec4-40ec-8457-20dae3018339] [Username=] [Password=] [KeepAlivePeriod=180] [CleanSession=True] [Timeout=00:03:00]
**[03:03:23.167]** [4] [MQTTnet.Core.Adapter.MqttChannelCommunicationAdapter] [Information]: **RX <<< MqttConnAckPacket**: [ConnectReturnCode=ConnectionAccepted] [IsSessionPresent=False]
[03:03:23.169] [3] [MQTTnet.Core.Client.MqttClient] [Information]: Received <<< MqttConnAckPacket: [ConnectReturnCode=ConnectionAccepted] [IsSessionPresent=False]
[03:03:23.172 - Client connected
[03:03:23.172 - Client: 001ea3ff-2ec4-40ec-8457-20dae3018339 connected in 10510 milliseconds.
[03:03:23.172 - Finished connecting 1 clients.
[03:03:23.174] [9] [MQTTnet.Core.Client.MqttClient] [Information]: Start sending keep alive packets.

**Comment by kpreisser** [comment]
Hi,

the issue with clients connecting slowly to the server (connecting 10 clients takes about 10 seconds to complete) is probably caused by this code in [`MqttServerAdapter.cs`](https://github.com/chkr1011/MQTTnet/blob/master/Frameworks/MQTTnet.NetStandard/Implementations/MqttServerAdapter.cs#L126) (line 126):
```c#
                finally
                {
                    //excessive CPU consumed if in endless loop of socket errors
                    await Task.Delay(TimeSpan.FromSeconds(1), cancellationToken).ConfigureAwait(false);
                }
```

Every time a client connects, the server waits 1 second before accepting the next client (regardless of whether the accepting of the previous client threw an exception). I think the intent of the code was probably to only wait when an exception occured.

However, I'm wondering if there is really a need to wait a bit when an exception occured to avoid an endless loop. Is there a case when `AcceptAsync()` throws, and all further calls throw also (except when the socket has been closed)?

Thanks!

**Comment by chkr1011** [comment]
@kpreisser Yes this was the intention in waiting. Doing this in the final block is just wrong. I will fix this.
@aronweiler There are some performance related changes pending. Probably your issue is fixed then. I will let you know when  a new test version is available.

**Comment by chkr1011** [comment]
@aronweiler I just released a previev of version 2.7.0 (nuget). Please let me know if this version works better for you.
Best regards
Christian
